lesson-15/src/index.js

```diff
@@ -0,0 +1,140 @@
+        id: product.id,
+        title: product.title
+      })
+      return acc
+    }, {})
+    console.log(groupedByCategory)
+  }
+  getAndGroupsProductByCategory()
+
+  // 7
+  async function fetchListOFProductsCountByCategories(category) {
+    const response = await fetch(`https://dummyjson.com/products/category/${category}`)
+    const listOfProducts = await response.json()
+    return listOfProducts.products.length
+  }
+
+  async function getAarrayOfCategories() {
+    const response = await fetch('https://dummyjson.com/products?limit=20&select=title,category')
+    const data = await response.json()
+    const arrayOfCategories = data.products.reduce((acc, product) => {
+      if (acc.includes(product.category)) {
+        return acc
+      } return [...acc, product.category]
+    }, [])
+    return arrayOfCategories
+  }
+
+  async function getProductByCategoryAndRerurnQuantity() {
+    const arrayOfCategories = await getAarrayOfCategories()
+    console.log(arrayOfCategories)
+    const productGroupsCountFetchArray = arrayOfCategories.map((category) => fetchListOFProductsCountByCategories(category))
+    console.log(productGroupsCountFetchArray)
+    const productsInSameGropsCount = await Promise.all(productGroupsCountFetchArray)
+    console.log('same products categories count', productsInSameGropsCount)
+    console.log(Object.fromEntries(
+      arrayOfCategories.map((tag, index) => [tag, productsInSameGropsCount[index]])
+    ))
+  }
+  getProductByCategoryAndRerurnQuantity()
+
+  // 8
+  async function getProductsByGroup(category) {
+    const response = await fetch(`https://dummyjson.com/products/category/${category}`)
+    const listOfProducts = await response.json()
+    return listOfProducts.products
+  }
+
+  async function fetchListOfProductsAndReturnPricesInGroup() {
+    const arrayOfCategories = await getAarrayOfCategories()
+    console.log(arrayOfCategories)
+    const productGroupsCountFetchArray = arrayOfCategories.map((category) => getProductsByGroup(category))
+    console.log(productGroupsCountFetchArray)
+    const productsInSameGrops = await Promise.all(productGroupsCountFetchArray)
+    // console.log('products sort by categories', productsInSameGrops)
+    const tileofproductsAndPrices = productsInSameGrops.map((arrayOfProducts) => {
+      const arrayOfTitles = arrayOfProducts.map((product) => product.title)
+      const arrayOfPrices = arrayOfProducts.map((product) => product.price)
+      return {
+        products: arrayOfTitles,
+        cheapestPrice: Math.min(...arrayOfPrices),
+        mostExpensivePrice: Math.max(...arrayOfPrices),
+        averagePrice: +(arrayOfPrices.reduce((acc, price) => acc + price, 0) / arrayOfPrices.length).toFixed(2)
+      }
+    })
+    console.log(tileofproductsAndPrices)
+    console.log(Object.fromEntries(
+      arrayOfCategories.map((category, index) => [category, tileofproductsAndPrices[index]])
+    ))
+  }
+  fetchListOfProductsAndReturnPricesInGroup()
```

Comment:
```suggestion
  async function getProductsByCategory(category) {
    const response = await fetch(`https://dummyjson.com/products/category/${category}`)
    const { products } = await response.json()

    return products
  }

  async function getCategoriesProductsInformation() {
    const categories = await getProductCategories()
    const productsByCategories = await Promise.all(categories.map((category) => (
      getProductsByCategory(category)
    )))

    return Object.fromEntries(
      categories.map((category, index) => {
        const products = productsByCategories[index]
        const prices = products.map(({ price }) => price)

        return [category, {
          products: products.map(({ title }) => title),
          cheapestPrice: Math.min(...prices),
          mostExpensivePrice: Math.max(...prices),
          averagePrice: (prices.reduce((acc, price) => acc + price, 0) / products.length).toFixed(2)
        }]
      })
    )
  }

  console.log(await getCategoriesProductsInformation())
```

зарефакторил по заметкам из предыдущих заданий